drivers/dma/dma_sam0.c

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright (c) 2018 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <soc.h>
+#include <dma.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(dma_sam0, CONFIG_DMA_LOG_LEVEL);
+
+#define DMA_REGS	((Dmac *)DT_ATMEL_SAM0_DMAC_0_BASE_ADDRESS)
+
+typedef void (*dma_callback)(void *callback_arg, u32_t channel,
+			     int error_code);
+
+struct dma_sam0_channel {
+	dma_callback cb;
+	void *cb_arg;
+};
+
+struct dma_sam0_data {
+	__aligned(16) DmacDescriptor descriptors[DMAC_CH_NUM];
+	__aligned(16) DmacDescriptor descriptors_wb[DMAC_CH_NUM];
+	struct dma_sam0_channel channels[DMAC_CH_NUM];
+};
+
+#define DEV_DATA(dev) \
+	((struct dma_sam0_data *const)(dev)->driver_data)
+
+
+/* Handles DMA interrupts and dispatches to the individual channel */
+static void dma_sam0_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct dma_sam0_data *data = DEV_DATA(dev);
+	struct dma_sam0_channel *chdata;
+	u16_t pend = DMA_REGS->INTPEND.reg;
+	u32_t channel;
+
+	/* Acknowledge all interrupts for the channel in pend */
+	DMA_REGS->INTPEND.reg = pend;
+
+	channel = (pend & DMAC_INTPEND_ID_Msk) >> DMAC_INTPEND_ID_Pos;
+	chdata = &data->channels[channel];
+
+	if (pend & DMAC_INTPEND_TERR) {
+		if (chdata->cb) {
+			chdata->cb(chdata->cb_arg, channel, -DMAC_INTPEND_TERR);
+		}
+	} else if (pend & DMAC_INTPEND_TCMPL) {
+		if (chdata->cb) {
+			chdata->cb(chdata->cb_arg, channel, 0);
+		}
+	}
+
+	/*
+	 * If more than one channel is pending, we'll just immediately
+	 * interrupt again and handle it through a different INTPEND value.
+	 */
+}
+
+/* Configure a channel */
+static int dma_sam0_config(struct device *dev, u32_t channel,
+			   struct dma_config *config)
+{
+	struct dma_sam0_data *data = DEV_DATA(dev);
+	DmacDescriptor *desc = &data->descriptors[channel];
+	struct dma_block_config *block = config->head_block;
+	struct dma_sam0_channel *channel_control;
+	DMAC_BTCTRL_Type btctrl = { .reg = 0 };
+	int key;
+
+	if (channel >= DMAC_CH_NUM) {
+		LOG_ERR("Unsupported channel");
+		return -EINVAL;
+	}
+
+	if (config->block_count > 1) {
+		LOG_ERR("Chained transfers not supported");
+		/* TODO: add support for chained transfers. */
+		return -ENOTSUP;
+	}
+
+	if (config->dma_slot >= DMAC_TRIG_NUM) {
+		LOG_ERR("Invalid trigger number");
+		return -EINVAL;
+	}
+
+	/* Lock and page in the channel configuration */
+	key = irq_lock();
+
+#ifdef DMAC_CHID_ID
```

Comment:
Since we are using a define from what I'm assuming is the HAL headers, can you add a comment either here or at the top of the file about it.

Reply:
Done, thanks.